torch/csrc/jit/type.h

```diff
@@ -408,8 +415,11 @@ struct TORCH_API ListType : public Type {
   static const TypeKind Kind = TypeKind::ListType;
 private:
   ListType(TypePtr elem)
-  : Type(TypeKind::ListType), elem(std::move(elem)) {}
+  : Type(TypeKind::ListType)
+  , elem(std::move(elem))
+  , has_free_variables_(getElementType()->hasFreeVariables()) {}
```

Comment:
Since the check is so simple, there's really no reason to cache it (the element type will cache it if it's complex).

Reply:
It is cached to avoid non-linear behavior. We start each (recursive) matching call with a check for hasFreeVariables().